### PR TITLE
feat(default_ad_api): add localization api 

### PR DESCRIPTION
**Pull request by isamu-takagi**
Signed-off-by: Takagi, Isamu <isamu.takagi@tier4.jp>

## Description

Introduce AD API for routing. This PR is the rviz helper, auto initial pose and AD API part of the diagram.
Here are the changes for each package:

autoware_ad_api_msgs: Add the localization API messages.
autoware_ad_api_specs: Add the localization API definitions.
component_interface_specs; Add the internal localization interface definitions.
default_ad_api: Add the localization API.
default_ad_api_helpers: Add the adaptor for RViz localization topics, and auto initialize node by GNSS.

block-diagram-initial-pose drawio

## Related links

https://github.com/autowarefoundation/autoware_launch/pull/72
https://github.com/autowarefoundation/autoware_launch/pull/70
https://github.com/tier4/tier4_autoware_msgs/pull/44
https://github.com/autowarefoundation/autoware.universe/pull/1500

~~https://github.com/autowarefoundation/autoware_launch/pull/71~~

## Tests performed

For planning simulation.
1. Merge all related PRs.
1. Launch planning simulation.
1. Check if the localization succeeds when initialpose from RViz.

For logging simulation.
1. Merge all related PRs.
1. Launch logging simulation (sample map).
1. Check if the localization succeeds automatically when the vehicle is stopped.
1. Disable auto initial pose (https://github.com/autowarefoundation/autoware_launch/pull/71)
   - Check if the localization succeeds when initialpose from RViz when vehicle is stopped.
   - Check if the localization failes when initialpose from RViz when vehicle is not stopped.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by isamu-takagi** [comment]
I'll split the PR.

**Comment by isamu-takagi** [comment]
This PR is ready.

File: system/default_ad_api_helpers/src/initial_pose_adaptor.cpp
Comment (kenji-miyake) [inline system/default_ad_api_helpers/src/initial_pose_adaptor.cpp:28]: [small question] Do you need this `template` word here?

File: system/default_ad_api_helpers/src/initial_pose_adaptor.hpp
Comment (kenji-miyake) [reply system/default_ad_api_helpers/src/initial_pose_adaptor.hpp:42]: Also, adding `const` would be polite?

File: system/default_ad_api_helpers/README.md
Comment (kenji-miyake) [inline system/default_ad_api_helpers/README.md:12]: Could you tell me where this service is implemented, please?

Comment (kenji-miyake) [reply system/default_ad_api_helpers/README.md:12]: I've found it in #1500.

File: common/autoware_ad_api_msgs/localization/srv/InitializeLocalization.srv
Comment (isamu-takagi) [reply common/autoware_ad_api_msgs/localization/srv/InitializeLocalization.srv:6]: | Error | Description |
| ---- | ---- |
| ERROR_UNSAFE | Requested when unsafe, such as while driving. |
| ERROR_GNSS_SUPPORT | GNSS is not supported. |
| ERROR_GNSS | GNSS is supported but an error related to GNSS occurred. |
| ERROR_ESTIMATION | An error occurred while pose estimation. |

File: system/default_ad_api_helpers/src/automatic_pose_initializer.cpp
Comment (isamu-takagi) [reply system/default_ad_api_helpers/src/automatic_pose_initializer.cpp:24]: Done https://github.com/autowarefoundation/autoware.universe/pull/1431/commits/6f51ee0408e203f63eb95298ae4cdc5e59af5430.

File: system/default_ad_api/src/localization.cpp
Comment (isamu-takagi) [reply system/default_ad_api/src/localization.cpp]: Done https://github.com/autowarefoundation/autoware.universe/pull/1431/commits/6f51ee0408e203f63eb95298ae4cdc5e59af5430.

File: system/default_ad_api/src/localization.hpp
Comment (isamu-takagi) [reply system/default_ad_api/src/localization.hpp]: Done https://github.com/autowarefoundation/autoware.universe/pull/1431/commits/6f51ee0408e203f63eb95298ae4cdc5e59af5430.

File: system/default_ad_api_helpers/README.md
Comment (isamu-takagi) [reply system/default_ad_api_helpers/README.md]: Split package https://github.com/autowarefoundation/autoware.universe/pull/1431/commits/7f6cfda8821b1e3ccee7219f0864ba87bdd6d467.

Comment (isamu-takagi) [reply system/default_ad_api_helpers/README.md]: Split package https://github.com/autowarefoundation/autoware.universe/commit/7f6cfda8821b1e3ccee7219f0864ba87bdd6d467.

File: system/default_ad_api_helpers/src/initial_pose_adaptor.hpp
Comment (isamu-takagi) [reply system/default_ad_api_helpers/src/initial_pose_adaptor.hpp]: Done https://github.com/autowarefoundation/autoware.universe/pull/1431/commits/f4ce8628b28b1f1e2ce76b4c2057876ac6c22cf1.

File: system/default_ad_api_helpers/src/initial_pose_adaptor.cpp
Comment (isamu-takagi) [reply system/default_ad_api_helpers/src/initial_pose_adaptor.cpp]: Replace main function https://github.com/autowarefoundation/autoware.universe/pull/1431/commits/7f6cfda8821b1e3ccee7219f0864ba87bdd6d467.

File: system/default_ad_api_helpers/src/automatic_pose_initializer.hpp
Comment (isamu-takagi) [reply system/default_ad_api_helpers/src/automatic_pose_initializer.hpp]: Done https://github.com/autowarefoundation/autoware.universe/pull/1431/commits/acdfc3f7a8c7073d5ebd043e8575ee550847e02f.

File: system/default_ad_api_helpers/src/automatic_pose_initializer.cpp
Comment (isamu-takagi) [reply system/default_ad_api_helpers/src/automatic_pose_initializer.cpp]: Done https://github.com/autowarefoundation/autoware.universe/pull/1431/commits/e5db1a7ae51dcbd5e4eaa121a352b6b96a1d0a6c.

File: system/default_ad_api_helpers/src/initial_pose_adaptor.cpp
Comment (isamu-takagi) [reply system/default_ad_api_helpers/src/initial_pose_adaptor.cpp]: It was unnecessary, removed  https://github.com/autowarefoundation/autoware.universe/pull/1431/commits/b94a6d807c31cfd619b020d31fd6d464d18c98d9. In the my first implementation, this function takes template argument `NodeT * node`, and template keyword is specified because the compiler doesn't know if the declare_parameter function is a normal function or a template function.

Comment (kenji-miyake) [reply system/default_ad_api_helpers/src/initial_pose_adaptor.cpp]: I see, thanks!

File: common/component_interface_specs/include/component_interface_specs/localization.hpp
Comment (kenji-miyake) [reply common/component_interface_specs/include/component_interface_specs/localization.hpp:35]: I suppose these are okay.

**Comment by kenji-miyake** [comment]
Waiting for #1494 to be merged.

File: system/default_ad_api_helpers/ad_api_adaptors/launch/rviz_adaptors.launch.xml
Comment (yukkysaito) [reply system/default_ad_api_helpers/ad_api_adaptors/launch/rviz_adaptors.launch.xml:7]: I agree. :pray:

Comment (isamu-takagi) [reply system/default_ad_api_helpers/ad_api_adaptors/launch/rviz_adaptors.launch.xml:7]: This adapter exposes `initialpose` for rviz, but not `fit_map_height`. It's just for internal use.

File: system/default_ad_api_helpers/ad_api_adaptors/README.md
Comment (isamu-takagi) [reply system/default_ad_api_helpers/ad_api_adaptors/README.md:12]: Same as https://github.com/autowarefoundation/autoware.universe/pull/1431#discussion_r958703883.

File: system/default_ad_api_helpers/ad_api_adaptors/launch/rviz_adaptors.launch.xml
Comment (isamu-takagi) [reply system/default_ad_api_helpers/ad_api_adaptors/launch/rviz_adaptors.launch.xml:6]: I won't change it this time. Considering the service, I think the name such as `~/sub/initial_pose` is clearer. So we should make guideline first.

File: system/default_ad_api_helpers/ad_api_adaptors/launch/rviz_adaptors.launch.xml
Comment (kenji-miyake) [reply system/default_ad_api_helpers/ad_api_adaptors/launch/rviz_adaptors.launch.xml:7]: I'm sorry, I couldn't understand well. How can others know it's for internal use? It looks the same as `/initialpose`.

Comment (isamu-takagi) [reply system/default_ad_api_helpers/ad_api_adaptors/launch/rviz_adaptors.launch.xml:7]: Anything not documented in the AD API is an internal implementation. From the point of view of AD API, initialpose is also treated as an internal implementation. It's just connected to rviz and localization in the same way that other component node is connected to multiple components.

Comment (kenji-miyake) [reply system/default_ad_api_helpers/ad_api_adaptors/launch/rviz_adaptors.launch.xml:7]: Please let me organize my question.

In [this comment](https://github.com/autowarefoundation/autoware.universe/pull/1431#discussion_r958703883), you said as if `initialpose` and `fit_map_height` are different.
My question was how others can understand the difference.

And my proposal is, by not exposing `fit_map_height` here, people can understand it.
Since `fit_map_height` is not used in RViz, I feel it's reasonable.

Comment (isamu-takagi) [reply system/default_ad_api_helpers/ad_api_adaptors/launch/rviz_adaptors.launch.xml:7]: I cannot understand what your problem is. It's common for a launch file to list both the interfaces a node provides and the interfaces it depends on. For example, [ad_service_state_monitor.launch.xml](https://github.com/autowarefoundation/autoware.universe/blob/main/system/ad_service_state_monitor/launch/ad_service_state_monitor.launch.xml) exposes `output/autoware_state` and `output/autoware_engage`, and uses implementation detail topic such as `input/vector_map`, `input/modified_goal`. What you are saying is the same as removing these.

Comment (kenji-miyake) [reply system/default_ad_api_helpers/ad_api_adaptors/launch/rviz_adaptors.launch.xml:7]: I feel it's strange that `initial_pose_adaptor` has the interface `fit_map_height`.
Ideally, I think as follows:
- `initial_pose_adaptor` only has the interface `initialpose`.
- When it receives an initialpose, it calls the service like `/localization/initialize_pose`.
- During `/localization/initialize_pose`, the localization component internally calls `fit_map_height`.
- So, `fit_map_height` won't be an interface of `initial_pose_adaptor`.

But since it's not so a big problem, if you don't agree with me and others don't have any comments, the current implementation is acceptable.

Comment (isamu-takagi) [reply system/default_ad_api_helpers/ad_api_adaptors/launch/rviz_adaptors.launch.xml:7]: Do you consider this to be implementation independent? It's different and this adaptor is the **default implementation**. In other words, this is the internal implementation that connects rviz with the API.

Comment (isamu-takagi) [reply system/default_ad_api_helpers/ad_api_adaptors/launch/rviz_adaptors.launch.xml:7]: >During /localization/initialize_pose, the localization component internally calls fit_map_height.
So, fit_map_height won't be an interface of initial_pose_adaptor.

It cannot. The API must trust the pose of `localization/initialize_pose`. The adapter modifies the height because it knows it's from rviz and is unreliable.

Comment (kenji-miyake) [reply system/default_ad_api_helpers/ad_api_adaptors/launch/rviz_adaptors.launch.xml:7]: I'm not sure about the default implementation, but I feel the localization component itself ideally should support 3D pose initialization.

If you call `fit_map_height` from `initial_pose_adaptor`, the issue will be hidden.
I mean, AD API can support 3D pose initialization, but the localization component only supports 2D pose initialization + height adjustment. Is that the proper component interface? :thinking:

Comment (isamu-takagi) [reply system/default_ad_api_helpers/ad_api_adaptors/launch/rviz_adaptors.launch.xml:7]: API and localization component only supports 3D pose initialization. This adaptor makes 3D pose from 2D pose initialization + height adjustment. Localization uses `fit_map_height` only for GNSS pose.

Comment (kenji-miyake) [reply system/default_ad_api_helpers/ad_api_adaptors/launch/rviz_adaptors.launch.xml:7]: Ah, I see. So you mean the height adjustment is RViz specific.
Please let me think it again.

File: system/default_ad_api_helpers/ad_api_adaptors/config/initial_pose.param.yaml
Comment (kenji-miyake) [inline system/default_ad_api_helpers/ad_api_adaptors/config/initial_pose.param.yaml:9]: [nits] Since it's 2D, the Z covariance would be extremely large. I'll leave it to you whether to apply this or not.

```suggestion
        0.0, 0.0, 1000000, 0.0,  0.0,  0.0,
```

File: system/default_ad_api_helpers/ad_api_adaptors/launch/rviz_adaptors.launch.xml
Comment (kenji-miyake) [reply system/default_ad_api_helpers/ad_api_adaptors/launch/rviz_adaptors.launch.xml:7]: Regarding the current implementation, I think it's acceptable, but feel a bit unnatural because the localization component has the interface `/localization/util/fit_map_height`.

So, let's proceed with the current implementation and revisit the interface in the future.
@isamu-takagi Is that okay for you?

File: system/default_ad_api_helpers/ad_api_adaptors/config/initial_pose.param.yaml
Comment (isamu-takagi) [reply system/default_ad_api_helpers/ad_api_adaptors/config/initial_pose.param.yaml:9]: I think that if the variance is too large it will not work as an initial position. In the current implementation, most of the particles will not converge to the correct position, as they will be scattered over a wide area up and down. So a lot of particles and computation time are required. We can trust Z because it is fitted to the map.

File: system/default_ad_api_helpers/ad_api_adaptors/launch/rviz_adaptors.launch.xml
Comment (isamu-takagi) [reply system/default_ad_api_helpers/ad_api_adaptors/launch/rviz_adaptors.launch.xml:7]: Alternatively, make `fit_map_height` a library and include it in each of GNSS (pose initializer) and RViz (API adaptor) node. It's a clean separation by design, but uses extra memory for the PCD map.

File: system/default_ad_api_helpers/ad_api_adaptors/config/initial_pose.param.yaml
Comment (kenji-miyake) [reply system/default_ad_api_helpers/ad_api_adaptors/config/initial_pose.param.yaml:9]: Hmm, is this the covariance after the map fitting? Then it's okay.

File: system/default_ad_api_helpers/ad_api_adaptors/launch/rviz_adaptors.launch.xml
Comment (kenji-miyake) [reply system/default_ad_api_helpers/ad_api_adaptors/launch/rviz_adaptors.launch.xml:7]: Sounds good, Thanks.
Let's reconsider at a later stage if necessary.

**Review by kenji-miyake — APPROVED** [review]

File: system/default_ad_api_helpers/automatic_pose_initializer/src/automatic_pose_initializer.cpp
Comment (kminoda) [inline system/default_ad_api_helpers/automatic_pose_initializer/src/automatic_pose_initializer.cpp:29]: As far as I tried on my PC, by setting the rate to higher value (e.g. `rclcpp::Rate(2.0)`), the `automatic_pose_initializer` started keep requesting for the pose initialization infinitely. It seems that the `timer_` was kept called, and this blocked `sub_state_` to update the `state_.state` from `UNINITIALIZED` to `INITIALIZED`.

Would you check if this happens in your PC, too? And if so, do you have any idea how to fix this issue?

Comment (kminoda) [reply system/default_ad_api_helpers/automatic_pose_initializer/src/automatic_pose_initializer.cpp:29]: @isamu-takagi  https://github.com/autowarefoundation/autoware.universe/issues/1750

**Pull request merged**
